### PR TITLE
Added seed argument to sample for localising reproducibility to a specific distribution

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References to other Issues or PRs

Fixes https://github.com/sympy/sympy/issues/20522

#### Other comments
If you want to keep this, then https://github.com/sympy/sympy/pull/20527 can be closed as this PR has the seeded version of the fix.
ping @oscarbenjamin @Upabjojr Please see if this fix is good. It was definitely a hard one, though after this the seed feature will be easy to extend.

#### Release Notes

* stats
    * `sympy.stats.sample` now has an optional `seed` argument.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
@maichmueller can you review this?

This provides the API to pass a seed in each call to sample or otherwise I think sample will use the library's global random state:
```
expr.sample(seed=1234)
```

It doesn't look like this makes it possible to seed once and then have multiple calls to sample use the same locally seeded state e.g. I would have expected something more like:
```
from sympy.stats import seed
seed(1234, library='scipy')
expr.sample()
expr.sample()
```

Also in the case that no seed is given does this create a local rng that persists across calls to sample and use that or is it always modifying the library's global state?

**Comment by czgdp1807** [comment]
The seeds provided are local to distribution. If nothing is given then library's seed settings are used.

**Comment by czgdp1807** [comment]
> from sympy.stats import seed
> seed(1234, library='scipy')
> expr.sample()
> expr.sample()

Well, this won't be possible to do in `1.7.1`.

**Comment by oscarbenjamin** [comment]
It might all be too much for 1.7.1 but we can certainly sort it out properly in 1.8.

My own suggestion for a simple fix in 1.7.1 was just that on the first call to sample with scipy the scipy random seed could be set using a random integer generated by the Python random module. That way the results from scipy are also seeded when the random module's seed is set. Maybe that's not worth doing though.

**Comment by oscarbenjamin** [comment]
Since SymPy didn't really support setting the seed before and there is a workaround which is to set the seed in scipy it might be best just to leave it for 1.7.1.

**Comment by czgdp1807** [comment]
Okay. I think this is the best thing that can be done without modifying the architecture too much considering the release of the previous version. 
I will open a new issue to discuss the API for seeding from scratch. When are we planning to release 1.8?

**Comment by czgdp1807** [comment]
In addition, let me know if this can be marked as `Stalled`, if it's not going to be merged.

**Comment by maichmueller** [comment]
I agree that adding a seeding option to the sampling iter is the best quick addition for now. A local state machine is the way to go in the longer run, but there are many details to be considered, e.g. how to store the local state so that all expression of randomness are with relation to it, and how to connect it with all supported libraries.

One more addition to the current setup could be a test for reproducibility, such as 
```python
seed = 0
n = 100
k = 2
samples = dict(scipy=[], numpy=[], pymc3=[])
for lib in samples.keys():
    for i in range(k):
        samples[lib].append(np.array(list(sample_iter(RV, numsamples=n, library=lib, seed=0))))
for lib, samps in samples.items():
    assert all([(samps[i] == samps[i + 1]).all() for i in range(k - 1)]), "Library " + lib + " samples differ despite seeding."
```

File: sympy/stats/drv.py [hunk 1]
```diff
@@ -69,20 +69,21 @@ def _pmf(self, x):
...
+    def __new__(cls, dist, size, seed=None):
+        return cls._sample_numpy(dist, size, seed)
 
     @classmethod
-    def _sample_numpy(cls, dist, size):
+    def _sample_numpy(cls, dist, size, seed):
         """Sample from NumPy."""
 
         import numpy
+        rand_state = numpy.random.RandomState(seed=seed)
```
Comment (maichmueller) [inline sympy/stats/drv.py:80]: This should be
```suggestion
        rand_state = numpy.random.default_rng(seed)
```
As numpy docs claim the default_rng (using PCG64) has better statistical properties than the mersenne twister of the RandomState.

File: sympy/stats/crv.py [hunk 2]
```diff
@@ -147,50 +147,94 @@ def __call__(self, *args):
...
+    def __new__(cls, dist, size, seed=None):
+        return cls._sample_numpy(dist, size, seed)
 
     @classmethod
-    def _sample_numpy(cls, dist, size):
+    def _sample_numpy(cls, dist, size, seed):
         """Sample from NumPy."""
 
         import numpy
+        rand_state = numpy.random.RandomState(seed=seed)
```
Comment (maichmueller) [inline sympy/stats/crv.py:221]: also here
```suggestion
        rand_state = numpy.random.default_rng(seed)
```

File: sympy/stats/joint_rv.py [hunk 3]
```diff
@@ -224,7 +231,11 @@ def _sample_pymc3(cls, dist, size):
 
         with pymc3.Model():
             pymc3_rv_map[dist.__class__.__name__](dist)
-            return pymc3.sample(size, chains=1, progressbar=False)[:]['X']
+            samples = pymc3.sample(size, chains=1, progressbar=False, random_seed=seed)[:]['X']
```
Comment (maichmueller) [inline sympy/stats/joint_rv.py:234]: Is this size querying related to adding the seed in pymc3? It looks like it fixes a different problem. If so, it might be better to change this as quick hotfix of a different PR.

File: sympy/stats/matrix_distributions.py [hunk 4]
```diff
@@ -66,40 +66,46 @@ def rv(symbol, cls, args):
...
         }
 
         dist_list = scipy_rv_map.keys()
 
         if dist.__class__.__name__ not in dist_list:
             return None
 
-        return scipy_rv_map[dist.__class__.__name__](dist, size)
+        samples = []
+        rand_state = numpy.random.RandomState(seed=seed)
```
Comment (maichmueller) [inline sympy/stats/matrix_distributions.py:93]:
```suggestion
        rand_state = numpy.random.default_rng(seed)
```

File: sympy/stats/matrix_distributions.py [hunk 5]
```diff
@@ -110,17 +116,23 @@ def _sample_numpy(cls, dist, size):
         if dist.__class__.__name__ not in dist_list:
             return None
 
-        return numpy_rv_map[dist.__class__.__name__](dist, size)
+        samples = []
+        import numpy
+        rand_state = numpy.random.RandomState(seed=seed)
```
Comment (maichmueller) [inline sympy/stats/matrix_distributions.py:121]:
```suggestion
        rand_state = numpy.random.default_rng(seed)
```

File: sympy/stats/matrix_distributions.py [hunk 6]
```diff
@@ -66,40 +66,46 @@ def rv(symbol, cls, args):
...
         if dist.__class__.__name__ not in dist_list:
             return None
 
-        return scipy_rv_map[dist.__class__.__name__](dist, size)
+        samples = []
+        rand_state = numpy.random.RandomState(seed=seed)
+        for _ in range(size[0]):
+            samp = scipy_rv_map[dist.__class__.__name__](dist, size[1], rand_state)
+            samples.append(samp)
+        return samples
```
Comment (maichmueller) [inline sympy/stats/matrix_distributions.py:97]: Does this fix something related to seeding or is it another issue?

File: sympy/stats/tests/test_continuous_rv.py [hunk 7]
```diff
@@ -333,6 +333,11 @@ def test_sample_continuous():
     sym, val = list(Z.pspace.sample().items())[0]
     assert sym == Z and val in Interval(0, oo)
 
+    samps = sample(Z, numsamples=3, seed=0)
+    assert round(next(samps), 4) == 0.7959
+    assert round(next(samps), 4) == 1.2559
+    assert round(next(samps), 4) == 0.9232
+
```
Comment (maichmueller) [inline sympy/stats/tests/test_continuous_rv.py:340]: Maybe add this to check all libraries, since it does not rely on hard-coded values (unsure if it fits in here)

```suggestion
    k = 2
    n = 10
    samples = dict(scipy=[], numpy=[], pymc3=[])
    for lib in samples.keys():
        for i in range(2):
            samples[lib].append(np.array(list(sample(Z, numsamples=n, library=lib, seed=0))))
    for lib, samps in samples.items():
        assert (
            all(
                [(samps[i] == samps[i + 1]).all() for i in range(k - 1)]
            ), 
            "Library " + lib + " samples differ despite seeding."
        )
    for lib in samples.keys():
        s0 = np.array(list(sample(Z, numsamples=n, library=lib, seed=0)))
        s1 = np.array(list(sample(Z, numsamples=n, library=lib, seed=1)))
        assert not (s0 == s1).all()
```

**Review by maichmueller — CHANGES_REQUESTED** [review]

**Comment by oscarbenjamin** [comment]
I guess we could merge this for now into 1.7.1 if we're sure that a per-call `seed=int` parameter is what we want in general. Otherwise we will have to defer any change here until 1.8.

@czgdp1807 @Upabjojr what do you think?

**Comment by oscarbenjamin** [comment]
Longer term I think that (as well as the seed=int) parameter there should be a way to seed all the sampling functions and set/get state functions. I can see a way to do this that does not seem that complicated:

SymPy stores a local state object for each sampling library. When imported these are just initialised to `None` so the random sampling state is:
```
sympy_sampling_state = {
    'random': None,
    'numpy': None,
    'scipy': None,
    'pymc3': None,
}
```
(these might actually be stored on the different class objects rather than in a single dict somewhere.)

SymPy provides a `seed_sampling` function so that `seed_sampling(1234)` turns the state into
```
sympy_sampling_state = {
    'random': random.Random(1234),
    'numpy': None,
    'scipy': None,
    'pymc3': None,
}
```

Now when there is a call to `sample(..., library='scipy', seed=789)` then if a seed is provided a random state `numpy.random.default_rng(789)` will be created and used internally for the sampling without otherwise affecting any state.

If there is a call to `sample(..., library='scipy')` and no seed is provided then the `sympy_sampling_state['scipy']` will be used. If the scipy state is at that time `None` then a new `numpy.random.default_rng(seed)` will be created (and stored in `sympy_sampling_state['scipy']`) with the seed generated as a random integer by `sympy_sampling_state['random']`.

What this means is that the user provided seed given to `seed_sampling` initialises a `Random` instance that generates seed for each of the other libraries. That way after `seed_sampling(1234)` the behaviour of all sampling from the other libraries is reproducible.

If at the time we want to generate the seed for scipy we find that `sympy_sampling_state['random']` is None (`seed_sampling` has not been called and no `seed=` argument supplied) then we do
```
    sympy_sampling_state['random'] = random.Random()
```
Then that is used to seed the rng for scipy which can be stored in `sympy_sampling_state['scipy']`.

Finally sympy can provide a `get_state` function that returns a dict like
```
state = {
    'random': sympy_sampling_state['random'].getstate(),
    'numpy': sympy_sampling_state['random'].get_state(),
    'scipy': None,
    'pymc3': None,
}
```
Here we call getstate for all libraries and put that into a dict with a `None` recorded for any library that wasn't used. Then there could also be a `set_state` function that takes the same dict and restores the state of all of the local RNGs.

This means that at most sympy draws a single sample from the random module in the instance where no seed is given either as an argument to `sample` or to the `seed_sampling` function. Otherwise It does not alter the global  random state at all and is unaffected by any other library or code that alters the global random state.

This also makes it possible to seed, set and get the state of all sampling libraries within a particular sympy version in a way that is reproducible when using the same version of sympy (I don't think we should try to guarantee reproducible behaviour between different sympy versions).

**Comment by czgdp1807** [comment]
> if we're sure that a per-call `seed=int` parameter is what we want in general.

I think this is a good addition and makes seed specific to an instance without interfering with other distributions. In fact, `scipy` also does something similar in `.rvs` API by accepting a `random_state` specific to an instance and same is the case with `pymc3` where a `random_seed` is accepted by it's sampling API. This also helps in avoiding the pain of managing seeds by `sympy` and all that part is done by the third party libraries. So, this is good overall.

@Upabjojr Your reviews will be helpful to decide on this. Thank you.

File: sympy/stats/joint_rv.py (hunk 3 above)
Comment (czgdp1807) [reply sympy/stats/joint_rv.py:234]: `size` is reflective of the size of each `sample`. The argument related to `seed` is `random_seed`.

File: sympy/stats/matrix_distributions.py (hunk 6 above)
Comment (czgdp1807) [reply sympy/stats/matrix_distributions.py:97]: This is for sampling of matrices from matrix distributions. It is related to this issue and should be there in this PR.

Comment (oscarbenjamin) [reply sympy/stats/matrix_distributions.py:97]: If this was previously just
```
return numpy_rv_map[dist.__class__.__name__](dist, size)
```
then can it not just be
```
return numpy_rv_map[dist.__class__.__name__](dist, size, state)
```
Why is the change to looping over `size[0]` needed?

**Comment by oscarbenjamin** [comment]
> In fact, `scipy` also does something similar in `.rvs` API by accepting a `random_state` specific to an instance and same is the case with `pymc3` where a `random_seed` is accepted by it's sampling API

Does this allow passing the seed as a `random_state` or `random_seed` or does it need to be an int?

**Review by maichmueller — APPROVED** [review]

**Comment by czgdp1807** [comment]
> Does this allow passing the seed as a `random_state` or `random_seed` or does it need to be an int?

`random_state` can be both an `int` and `numpy.random.RandomState`/`numpy.random.default_rng`(https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.rvs.html).
`pymc3` required `random_seed` to be an `int`/list of `int` objects. (https://docs.pymc.io/api/inference.html).

Since, `int` type is common to all the currently supported libraries, `sympy.stats.sample` supports `int` type for seed. This also helps to easily manage seed forwarding to different libraries depending on how they accept it, since, `int` seed can be used to local generators for any library.

**Comment by oscarbenjamin** [comment]
@Upabjojr what do you think of this?

This is now the only release blocker for 1.7.1. Currently the only change in 1.7.1 is #20521 which fixes sampling to use the correct scipy APIs for common distributions. No other issues with 1.7 have been reported.

We could merge this, release 1.7.1 and apply this change to master as well.

Or otherwise we could postpone any change here until 1.8.

If we are not 100% sure that we want this API change then we should postpone because 1.7.1 is a bugfix release.

**Comment by Upabjojr** [comment]
> @Upabjojr Your reviews will be helpful to decide on this. Thank you.

Maybe we should just add `**kwargs` to `sample`, letting users pass custom arguments to the sampling backend. Easiest design, easiest way.

Personally, I wouldn't change anything nor add any support for seeding in SymPy. SymPy users can still set the seed by calling the external library and setting the seed manually.

**Comment by Upabjojr** [comment]
If you really like to add a seeding parameter to `sample( ... )`, then go for this PR. No objection on my side.

**Comment by oscarbenjamin** [comment]
Thanks for the review.

Okay, then I'll merge this. @czgdp1807 can you open a similar PR for master?

**Comment by czgdp1807** [comment]
> Maybe we should just add `**kwargs` to `sample`, letting users pass custom arguments to the sampling backend. Easiest design, easiest way.

The reason why we cannot do this is that, sometimes libraries like `numpy`, `cupy`* don't accept arguments for accepting seeds specific to a distribution. We need to create objects like `RandomState` or `default_rng` to sample distributions. Now, `**kwargs` can be helpful with some libraries but for all, we need an `int` parameter. In addition, `**kwargs` will make API ambiguous.



> Personally, I wouldn't change anything nor add any support for seeding in SymPy. SymPy users can still set the seed by calling the external library and setting the seed manually.

I also agree with this. But the current change doesn't introduce burden of seed management on SymPy, it is just forwarding the seed to the third party library. In other words, SymPy is just acting like a mediator between user and the third libraries for this task.

**Comment by czgdp1807** [comment]
>  @czgdp1807 can you open a similar PR for master?

I can directly make changes to @Upabjojr 's branch of https://github.com/sympy/sympy/pull/20494 if he is okay with this. This way we can make architectural changes and add seed parameter in a single PR.

**Comment by Upabjojr** [comment]
> In addition, `**kwargs` will make API ambiguous.

It's called matplotlib-style :)

> I can directly make changes to @Upabjojr 's branch of #20494 if he is okay with this. This way we can make architectural changes and add seed parameter in a single PR.

Now that we have seeds, I was thinking about reintroducing classes in that PR. A class would avoid the need of passing the `seed` argument on every method, though it looks a bit complicated (unless using Python 3.8): 
https://stackoverflow.com/questions/24063788/python3-singledispatch-in-class-how-to-dispatch-self-type

**Comment by oscarbenjamin** [comment]
> A class would avoid the need of passing the `seed` argument on every method

You'd still need to pass the seed in every call because the whole point of the API introduced here is that you can pass the seed explicitly when calling `sample`.

**Comment by Upabjojr** [comment]
> > A class would avoid the need of passing the `seed` argument on every method
> 
> You'd still need to pass the seed in every call because the whole point of the API introduced here is that you can pass the seed explicitly when calling `sample`.

Of course. But this way you don't need to add a seed parameter to every distribution dispatched call.

**Comment by czgdp1807** [comment]
> It's called matplotlib-style :)

I personally quite struggle with `matplotlib`. In fact, I couldn't survive plotting using `matplotlib` without their docs (which is also hard to read, due to a lot of content in a single page).

Just a cautious note that now since `seed` is introduced, we shouldn't remove it. In fact, creating an internal local state inside `sympy.stats` will definitely not be a good idea. Right now, the seeds aren't shared between two distributions and hence, in a large project using `sympy.stats` developers will not have to worry about in other parts of code since, seed is local to a distribution and not shared among a bunch of distributions. We should make sure any change we make, this property of locality and user facing API aren't changed.

**Comment by maichmueller** [comment]
> Just a cautious note that now since seed is introduced, we shouldn't remove it. In fact, creating an internal local state inside sympy.stats will definitely not be a good idea

I disagree, these should not be mutually exclusive. The current API merely allows to create a single reproducible sampling call of some distribution. However, afaik it still does not allow to create a reproducible sample chain, which is crucial if sympy is to be used for e.g. Bayesian statistics.
One solution is to pass a numpy Generator as `seed`, so that the rng machine is client side, but we would need to figure out how to combine this with pymc3. If pymc3 ends up being incompatible with a numpy generator, then the only route forward would be a local state in sympy imho.

**Comment by czgdp1807** [comment]
> The neatest solution would be to pass a numpy Generator as seed, so that the rng machine is client sided

Is the following what you are looking for,

```python
In [1]: from sympy.stats import *

In [2]: N = Normal('N', 0, 1)

In [3]: import numpy

In [4]: rng = numpy.random.default_rng(0)

In [5]: samples = sample(N, numsamples=10, seed=rng)
/home/czgdp1807ssd/sympy_project/sympy/sympy/stats/rv.py:1099: UserWarning: 
The return type of sample has been changed to return an iterator
object since version 1.7. For more information see
https://github.com/sympy/sympy/issues/19061
  warnings.warn(filldedent(message))

In [6]: list(samples)
Out[6]: 
[0.1257302210933933, -0.1321048632913019, 0.6404226504432821, 0.10490011715303
971, -0.535669373161111, 0.36159505490948474, 1.3040000451301372, 0.9470809631
292422, -0.7037352358069926, -1.2654214710460525]
```

**Comment by maichmueller** [comment]
That is precisely what I mean. I was expecting this to work with scipy and numpy, but not with pymc3 (correct?). The question then is, how to address this with pymc3 if not with a local state?
